### ISSUE TITLE
Bug: Deadlock in concurrent update()s

### ISSUE THREAD

**Issue by chun0nick** [unum-cloud/usearch#396]
### Describe the bug

Deadlock came up during testing, appears to be an issue in index.hpp. This testing involved deleting a bunch of vectors and then, with multiple threads, calling add(), which then called update() on the deleted nodes.

In `update`: we acquire the lock on the deleted node [here](https://github.com/unum-cloud/usearch/blob/main/include/usearch/index.hpp#L2422). We then call `connect_node_across_levels_` which invokes `search_for_one_` and `search_to_insert_`, while [locking](https://github.com/unum-cloud/usearch/blob/main/include/usearch/index.hpp#L3375) and [locking](https://github.com/unum-cloud/usearch/blob/main/include/usearch/index.hpp#L3441) (respectively) every candidate node it looks at.

This first lock on the deleted node is the problem. If another thread performing an update selects that deleted node as a "next" candidate, we have to wait until the other thread is completely done updating, as we don't drop the lock until the end of the call. What we've seen somewhat consistently (with 16 threads updating) is a deadlock, where each of the threads needs the lock on a candidate that another thread updating already holds, and that thread in turn needs the lock on a candidate a different thread holds.

This may be the issue described in https://github.com/unum-cloud/usearch/issues/354?

### Steps to reproduce

Create an index, add a reasonable number of vectors. Delete them - then concurrently call add (update) with a bunch of threads.

### Expected behavior

The concurrent updates should not conflict with each other.

### USearch version

v2.8.16

### Operating System

Amazon Linux 2

### Hardware architecture

x86

### Which interface are you using?

Other bindings

### Contact Details

chun0nick@gmail.com

### Is there an existing issue for this?

- [X] I have searched the existing issues

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by ashvardanian** [comment]
Good catch! Looking into it! Feel free to submit a test-case or patches via PR while I do 🤗

**Comment by ashvardanian** [comment]
Replicated it, will make part of the primary testing suite in C++. Thanks for finding and reporting the issue, @chun0nick! Would you be open to be mentioned as a contributor for this patch? 🤗

**Comment by chun0nick** [comment]
Awesomely fast response! And sorry - was planning to give you a test case today. Sure, would be happy to be mentioned as a contributor.

**Comment by sef43** [comment]
I had what looked to be a deadlock happen when using the Python api (as part of usearch_molecules indexing step) when using 32 threads. Could this bug be the cause of that?

**Comment by ashvardanian** [comment]
Yes, @sef43, the issue is coming from the core in update-heavy workloads. Haven't finished patching yet, was busy the last few days with new bindings for [UForm](https://github.com/unum-cloud/uform) and new algos in [StringZilla](https://github.com/ashvardanian/stringzilla).

**Comment by chun0nick** [comment]
Hi, was curious if there was movement on this issue - I've taken a few cracks at a fix but have quickly found myself touching quite a bit of code. Thanks!

**Comment by ashvardanian** [comment]
I've spent some time on this recently as well. Concurrent updates are quite tricky, especially if you want to avoid inefficient memory usage. I found no clean way of keeping a 2-layer solution, separating the logic between index_gt and index_dense_gt. 

A better solution might be to have a more advanced compare-exchange-like operation in index_gt. 

After my weekend patches, every other core C++ test on the main-dev branch passes, except for these concurrent updates and weird SEGFAULT in Python bindings, so its a core priority 🤗

**Comment by vitarb** [comment]
@ashvardanian it's been a while, I wonder if it fell out of your radar, were you able to get this fixed?

**Comment by ashvardanian** [comment]
Hi @vitarb! I would heavily discourage concurrent removals and updates in the current v2. I'd love to learn more about your usecase and if you'd like to be participating in the v3 preview. We can continue over email as well - ash.vardanian@unum.cloud 🤗
